### ISSUE TITLE
disorder `SubscribeMessages` event

### ISSUE THREAD

**Issue by akinazuki** [tmaxmax/go-sse#18]
### Here is my testcase
```go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"strings"

	"github.com/tmaxmax/go-sse"
)

type OpenAIRequest struct {
	Model    string    `json:"model"`
	Messages []Message `json:"messages"`
	Stream   bool      `json:"stream"`
}
type Message struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}
type OpenAIResponse struct {
	OrderId           int
	ID                string `json:"id"`
	Object            string `json:"object"`
	Created           int    `json:"created"`
	Model             string `json:"model"`
	SystemFingerprint string `json:"system_fingerprint"`
	Choices           []struct {
		Index int `json:"index"`
		Delta struct {
			Content string `json:"content"`
		} `json:"delta"`
		FinishReason interface{} `json:"finish_reason"`
	} `json:"choices"`
}

func main() {
	msg := make(chan string)
	go requestOpenAI([]Message{
		{
			Role:    "user",
			Content: "Hello",
		},
	}, msg)
	finalMsg := ""
	for m := range msg {
		finalMsg += m
	}
	fmt.Println(finalMsg)
}

func requestOpenAI(msgs []Message, events chan string) {
	OpenAIKey := "sk-xxx"
	defer close(events)
	prompt, _ := json.Marshal(OpenAIRequest{
		Model:    "gpt-4-1106-preview",
		Messages: msgs,
		Stream:   true,
	})
	r, _ := http.NewRequest(http.MethodPost, "https://api.openai.com/v1/chat/completions", strings.NewReader(string(prompt)))
	r.Header.Add("Authorization", fmt.Sprintf("Bearer %s", OpenAIKey))
	r.Header.Add("Content-Type", "application/json")
	conn := sse.NewConnection(r)

	orderId := 0
	conn.SubscribeMessages(func(ev sse.Event) {
		orderId++
		var resp OpenAIResponse
		json.Unmarshal([]byte(ev.Data), &resp)
		resp.OrderId = orderId
		if len(resp.Choices) == 0 || resp.Choices[0].FinishReason == "stop" {
			return
		}
		fmt.Printf("Got event, evtId: %d, msg: %s\n", orderId, resp.Choices[0].Delta.Content)
		events <- resp.Choices[0].Delta.Content
	})

	if err := conn.Connect(); err != nil {
		log.Println(err)
	}
}
```

Run output:

```
go run test.go
Got event, evtId: 1, msg: 
Got event, evtId: 3, msg: Hello
Got event, evtId: 3, msg:  How
Got event, evtId: 4, msg: !
Got event, evtId: 5, msg:  assist
Got event, evtId: 6, msg:  can
Got event, evtId: 7, msg:  I
Got event, evtId: 8, msg: ?
Got event, evtId: 9, msg:  you
Got event, evtId: 10, msg:  today
Hello How! assist can I? you today
```

**Comment by tmaxmax** [comment]
Hi there! Please try this to fix your issue:
```go
func requestOpenAI(msgs []Message, events chan string) {
    // keep the same code

    sseEvents := make(chan sse.Event)
    conn.SubscribeMessages(func (ev sse.Event) {
        sseEvents <- ev
    })

    done := make(chan struct{})
    go func() {
        defer close(done)
        orderID := 0
        for ev := range sseEvents {
            // process the event the same way as you did
            events <- resp.Choices[0].Delta.Content
        }
    }()

    if err := conn.Connect(); err != nil {
        log.Println(err)
    }
    
    <-done
}
```

Anyway, I think the problem occurs because under the hood the callback passed to `SubscribeMessages` is called in a separate goroutine for each event (and as a fun fact, because of this a data race occurs in your code when you update `orderId` inside the callback). Given that you process the data in the callback, the messages get desynchronized.

In hindsight, the decision to have the callbacks run concurrently (or to use callbacks instead of channels, now that I think about it) wasn't the best one, especially because it's hidden away from the library user. The client requires a rewrite, as there are other problems with it – for example #17; tests and testability also need to be improved. This footgun will be fixed in a future version, as I'll change the API to use channels rather than callbacks, among other changes.

Until then, please let me know my proposed solution solves the issue for you! Thanks for flagging this, will probably be useful to other people, too.

**Comment by akinazuki** [comment]
thanks for your reply, but the problem isn't solved, the response is still disordered

here is my modified code:


```go
func requestOpenAI(msgs []Message, events chan string) {
	OpenAIKey := os.Getenv("OPENAI_KEY")
	defer close(events)
	prompt, _ := json.Marshal(OpenAIRequest{
		Model:    "gpt-4-1106-preview",
		Messages: msgs,
		Stream:   true,
	})
	r, _ := http.NewRequest(http.MethodPost, "https://api.openai.com/v1/chat/completions", strings.NewReader(string(prompt)))
	r.Header.Add("Authorization", fmt.Sprintf("Bearer %s", OpenAIKey))
	r.Header.Add("Content-Type", "application/json")
	conn := sse.NewConnection(r)

	sseEvents := make(chan sse.Event)
	conn.SubscribeMessages(func(ev sse.Event) {
		sseEvents <- ev
	})

	go func() {
		defer close(events)
		orderId := 0
		for ev := range sseEvents {
			orderId++
			var resp OpenAIResponse
			json.Unmarshal([]byte(ev.Data), &resp)
			resp.OrderId = orderId
			if len(resp.Choices) == 0 || resp.Choices[0].FinishReason == "stop" {
				return
			}
			fmt.Printf("Got event, evtId: %d, msg: %s\n", orderId, resp.Choices[0].Delta.Content)
			events <- resp.Choices[0].Delta.Content
		}
	}()

	if err := conn.Connect(); err != nil {
		log.Println(err)
	}
}
```

```
 go run test.go
Got event, evtId: 1, msg: 
Got event, evtId: 2, msg:  How
Got event, evtId: 3, msg: Hello
Got event, evtId: 4, msg: !
Got event, evtId: 5, msg:  assist
Got event, evtId: 6, msg:  can
Got event, evtId: 7, msg:  I
 HowHello! assist can I
```

**Comment by tmaxmax** [comment]
I see. Can you try to get the raw SSE stream response? Using `curl` inside bash you could execute the following command:
```
$ curl -X POST -N 'https://api.openai.com/v1/chat/completions' \
  -H "Authorization: Bearer $OPENAI_KEY" \
  -H "Content-Type: application/json" \
  -d '{"model":"gpt-4-1106-preview","messages":[{"role":"user","content":"hello"}],"stream":true}'
```
Run it and paste the output here, after it's finished. Didn't test the command, you should make sure yourself that the input is correct.

**Comment by akinazuki** [comment]
```
data: {"id":"chatcmpl-8Mc3MSMWA3e3BrTPZN6losXi9ZoeC","object":"chat.completion.chunk","created":1700400184,"model":"gpt-4-1106-preview","system_fingerprint":"fp_a24b4d720c","choices":[{"index":0,"delta":{"role":"assistant","content":""},"finish_reason":null}]}

data: {"id":"chatcmpl-8Mc3MSMWA3e3BrTPZN6losXi9ZoeC","object":"chat.completion.chunk","created":1700400184,"model":"gpt-4-1106-preview","system_fingerprint":"fp_a24b4d720c","choices":[{"index":0,"delta":{"content":"Hello"},"finish_reason":null}]}

data: {"id":"chatcmpl-8Mc3MSMWA3e3BrTPZN6losXi9ZoeC","object":"chat.completion.chunk","created":1700400184,"model":"gpt-4-1106-preview","system_fingerprint":"fp_a24b4d720c","choices":[{"index":0,"delta":{"content":"!"},"finish_reason":null}]}

data: {"id":"chatcmpl-8Mc3MSMWA3e3BrTPZN6losXi9ZoeC","object":"chat.completion.chunk","created":1700400184,"model":"gpt-4-1106-preview","system_fingerprint":"fp_a24b4d720c","choices":[{"index":0,"delta":{"content":" How"},"finish_reason":null}]}

data: {"id":"chatcmpl-8Mc3MSMWA3e3BrTPZN6losXi9ZoeC","object":"chat.completion.chunk","created":1700400184,"model":"gpt-4-1106-preview","system_fingerprint":"fp_a24b4d720c","choices":[{"index":0,"delta":{"content":" can"},"finish_reason":null}]}

data: {"id":"chatcmpl-8Mc3MSMWA3e3BrTPZN6losXi9ZoeC","object":"chat.completion.chunk","created":1700400184,"model":"gpt-4-1106-preview","system_fingerprint":"fp_a24b4d720c","choices":[{"index":0,"delta":{"content":" I"},"finish_reason":null}]}

data: {"id":"chatcmpl-8Mc3MSMWA3e3BrTPZN6losXi9ZoeC","object":"chat.completion.chunk","created":1700400184,"model":"gpt-4-1106-preview","system_fingerprint":"fp_a24b4d720c","choices":[{"index":0,"delta":{"content":" assist"},"finish_reason":null}]}

data: {"id":"chatcmpl-8Mc3MSMWA3e3BrTPZN6losXi9ZoeC","object":"chat.completion.chunk","created":1700400184,"model":"gpt-4-1106-preview","system_fingerprint":"fp_a24b4d720c","choices":[{"index":0,"delta":{"content":" you"},"finish_reason":null}]}

data: {"id":"chatcmpl-8Mc3MSMWA3e3BrTPZN6losXi9ZoeC","object":"chat.completion.chunk","created":1700400184,"model":"gpt-4-1106-preview","system_fingerprint":"fp_a24b4d720c","choices":[{"index":0,"delta":{"content":" today"},"finish_reason":null}]}

data: {"id":"chatcmpl-8Mc3MSMWA3e3BrTPZN6losXi9ZoeC","object":"chat.completion.chunk","created":1700400184,"model":"gpt-4-1106-preview","system_fingerprint":"fp_a24b4d720c","choices":[{"index":0,"delta":{"content":"?"},"finish_reason":null}]}

data: {"id":"chatcmpl-8Mc3MSMWA3e3BrTPZN6losXi9ZoeC","object":"chat.completion.chunk","created":1700400184,"model":"gpt-4-1106-preview","system_fingerprint":"fp_a24b4d720c","choices":[{"index":0,"delta":{},"finish_reason":"stop"}]}

data: [DONE]
```

**Comment by tmaxmax** [comment]
Thank you. The input is obviously not the issue, just wanted to make sure. I'll try and reproduce this locally with a minimal example and get back to you.

**Comment by tmaxmax** [comment]
Should be fixed in release v0.7.0!
